Codify breaking change guidelines (#13066)
solana-labs/solana: docs/src/apps/backwards-compatibility.md

---
title: Backwards Compatibility Policy
---

As the Solana developer ecosystem grows, so does the need for clear expectations around
breaking API and behavior changes affecting applications and tooling built for Solana.
In a perfect world, Solana development could continue at a very fast pace without ever
causing issues for existing developers. However, some compromises will need to be made
and so this document attempts to clarify and codify the process for new releases.

### Expectations

> - Solana APIs, SDKs, and CLI tooling are tied to validator software releases (with a few

Reviewer: The monorepo release is also morphing into the program SDK release as well, it's not really a validator software release anymore.  The "Solana software release"?

  (with a few [exceptions](#exceptions)).

> - Software for a `MINOR` version release will be compatible across all software on the

Reviewer 2: ```suggestion
- Software for a `MINOR` version release will be compatible across all software back to the
```

Reviewer 2: Or "since"

>   **same** `MINOR` version _regardless_ of `PATCH` version.

Reviewer: I think we should aim to stay compatible, at least, to the first `PATCH` release of the previous `MINOR` version.

Author: Sounds good, then any deprecated feature will be deprecated for a full MINOR release before being removed.

Reviewer 2: I assume that once a feature has been deprecated in a MINOR version, the next MINOR version does not need to remain compatible with it, correct? If so, I think we should make that explicit. (If not, I don't follow how we ever remove things.)

Also, I think it is worth keeping the verbiage that says that the software will remain compatible across all PATCH releases in the same MINOR release.

Reviewer: > I assume that once a feature has been deprecated in a MINOR version, the next MINOR version does not need to remain compatible with it, correct? If so, I think we should make that explicit. (If not, I don't follow how we ever remove things.)

Fully compatible back to at least two MINOR.0 versions in the past.   So the tip of master (1.5.0) should be fully compatible all the way back to 1.3.0+.   When we move to 1.6.0, at our option we can start removing features first deprecated in the 1.3.x release if/when they get in the way.

Author: Why at least two versions in the past? I think that one is sufficient so that things deprecated in 1.3.x can be removed in 1.5.0. At the current rate, a single MINOR release cycle is 2 months right? That means a change will be deprecated for 4-6 months if we support two versions rather than 2-4 months on one. The difference of 2 months just pushes back the inevitable for developers, they can expect breaking changes every MINOR.0 anyways.

Author: @mvines ping! I'm pushing back on the "at least 2 version" to just 1 version. Thoughts?

Reviewer 3: I think the idea was to communicate that any deprecation will live on for _at least_ one full release cycle.  So we won't deprecate in the last patch of 1.4 and remove in 1.5.0

Reviewer: I think we can expect at least some external devs to upgrade slowly, breaking them every 2-4 months is probably too frequent.  

> I think that one is sufficient so that things deprecated in 1.3.x can be removed in 1.5.0.
 
It's fair I think to mark things as deprecated the 1.3.x release while it remains on the stable channel.    This means what I marked as deprecated in 1.3.x last week and be removed from master today, and we'd then lose to ability to validate master against Serium.   From their perspective, we'd be asking them to remove a feature that's been deprecated for a week or less

Author: > I think we can expect at least some external devs to upgrade slowly, breaking them every 2-4 months is probably too frequent.

Hmm, how about breaking changes only on even numbered MINOR releases?

> This means what I marked as deprecated in 1.3.x last week could be removed from master today

Ah, putting it this way is interesting. Does that imply that we can't make breaking changes to beta?

Reviewer: > Hmm, how about breaking changes only on even numbered MINOR releases?

Breaking changes are a major release bump?   Like what semver wants us to do anyway :)

> Does that imply that we can't make breaking changes to beta?

💯.

Author: > Breaking changes are a major release bump? Like what semver wants us to do anyway :)

Haha, that works!

Author: Alright, docs have been updated to promise proper semver 👍

- Solana **DOES NOT** guarantee that software on _non-sequential_ `MINOR` version releases
  will always be compatible. _(For example, 1.3.x will likely not be compatible with 1.5.x)_
- Solana software releases **DO NOT** strictly follow semantic versioning, details below.

### Release Cadence

The Solana RPC API, Rust SDK, CLI tooling, and BPF Program SDK are all updated and shipped
along with each Solana software release and should always be compatible between patch
updates of a particular minor version release.

#### Minor Releases (1.x.0)

Experimental changes and new proposal implementations are added to new `MINOR` version
releases (ex. 1.4.0) and are first run on Solana's Tour de SOL testnet cluster. After
those changes have proven to be stable, the Mainnet Beta cluster will be updated to the

> new minor version.

Reviewer 3: ```suggestion
new `MINOR` version.
```

#### Patch Releases (1.0.x)

Low risk features, non-breaking changes, and security and bug fixes are shipped as part
of `PATCH` version releases (ex. 1.0.11).

### RPC API

Patch releases:
- New RPC endpoints and features
- Bug fixes
- Security fixes

Minor releases:
- Endpoint / feature deprecation

> - Removal of previous minor version deprecated features

Reviewer 3: ```suggestion
- Removal of previous `MINOR` version deprecated features
```

### Program SDK

Patch releases:
- New APIs
- Bug fixes
- Performance improvements
- Security fixes

Minor releases:
- Removal of deprecated APIs
- Backward compatibility breaking changes

### CLI Tools

Patch releases:
- New subcommands
- Bug and security fixes
- Performance improvements

Minor releases:
- Switch to new RPC API endpoints / configuration introduced in the previous minor version.
- Subcommand / argument deprecation
- Removal of previous minor version deprecated features

### Runtime Features

New Solana runtime features are feature-switched and manually activated. Runtime features
include: the introduction of new native programs, sysvars, and syscalls; and changes to

> their behavior.

Reviewer 3: ```suggestion
their behavior. Feature activation is cluster agnostic, allowing confidence to be built on
Testnet before activation on Mainnet-beta.
```

The release process is as follows:

1. New runtime feature is included in a new release, deactivated by default
2. Once sufficient staked validators upgrade to the new release, the runtime feature switch is activated manually with an instruction
3. The feature takes effect at the beginning of the next epoch

### Infrastructure Changes

#### Public API Nodes

Solana provides publicly available RPC API nodes for all developers to use. The Solana team
will make their best effort to communicate any changes to the host, port, rate-limiting behavior,
availability, etc. However, we recommend that developers rely on their own validator nodes to

> discourage centralized usage of Solana operated nodes.

Reviewer 3: ```suggestion
discourage dependence upon Solana operated nodes.
```

#### Local cluster scripts and Docker images

> Breaking changes will be limited to minor version updates. Patch updates should always

Reviewer 3: ```suggestion
Breaking changes will be limited to `MINOR` version updates. `PATCH` updates should always
```

### CLI Tools

Patch releases:
- Bug and security fixes
- Performance improvements
- Subcommand / argument deprecation

Minor releases:
- New subcommands

Major releases:
- Switch to new RPC API endpoints / configuration introduced in the previous major version.
- Removal of deprecated features

### Runtime Features

New Solana runtime features are feature-switched and manually activated. Runtime features
include: the introduction of new native programs, sysvars, and syscalls; and changes to
their behavior. Feature activation is cluster agnostic, allowing confidence to be built on
Testnet before activation on Mainnet-beta.

The release process is as follows:

1. New runtime feature is included in a new release, deactivated by default
2. Once sufficient staked validators upgrade to the new release, the runtime feature switch
  is activated manually with an instruction
3. The feature takes effect at the beginning of the next epoch

### Infrastructure Changes

#### Public API Nodes

Solana provides publicly available RPC API nodes for all developers to use. The Solana team
will make their best effort to communicate any changes to the host, port, rate-limiting behavior,

> availability, etc. However, we recommend that developers rely on their own validator nodes to

Reviewer 2: To this point, should we include rate-limiting info on docs.solana.com/clusters with the RPC URL?
